### PR TITLE
Meta device initialization for FSDP in Fabric

### LINKED ISSUE #16448: Adopt `FakeTensorMode` for FSDP
### Description & Motivation

Currently, very large models can't be instantiated before sharding if they don't fit in CPU memory. The solution to this, and similar to how other libraries like deepspeed do it, is to create fake tensors that don't allocate memory, then shard them, and only materialze them once sharded on each device. 

### Pitch

Adopt the [FakeTensorMode](https://github.com/pytorch/pytorch/blob/master/torch/_subclasses/fake_tensor.py#L721) context manager. It is not officially documented, but we can start experimenting it in Fabric.

The usage in PyTorch would be:

```py
with FakeTensorMode():
    model = MyModel(...)  # tensors are fake and don't allocate memory
```

This would translate to 
```py
with fabric.sharded_model():
    model = MyModel(...)
```

in Fabric.

### Alternatives

A similar mechanism exists in [torchdistx](https://github.com/pytorch/torchdistx/blob/2071f8b4ae2325487f5e555f48fc21c9161aae98/src/python/torchdistx/deferred_init.py).

### Additional context

Proposed by @justusschock 

_No response_

cc @borda @carmocca @justusschock @awaelchli

### LINKED ISSUE #18008: Multi-node training with FSDP results in weird behaviour
### Bug description

Here is my train script for llama 65b
```import argparse
import os
import sys
import time
from functools import partial
from pathlib import Path

import lightning as L
import torch
from lightning.fabric.strategies import FSDPStrategy
from prepare import (calculate_training_steps, get_batch, load_datasets,
                     prepare, validate)
from torch.distributed.fsdp.wrap import transformer_auto_wrap_policy, size_based_auto_wrap_policy
from torch.nn.utils.init import skip_init
from torch.distributed.fsdp import (
   FullyShardedDataParallel,
   CPUOffload,
)

# support running without installing as a package
wd = Path(__file__).parent.parent.resolve()
sys.path.append(str(wd))

from lit_llama.model import Block, LLaMA, LLaMAConfig
from lit_llama.utils import save_model_checkpoint

# compilation fails as it does not support torch.complex64 for RoPE
# compile = False

beta1 = 0.9
beta2 = 0.95
grad_clip = 1.0



def train(
    output_dir: str,
    data_dir: str,
    eval_interval: int = 2000,
    eval_iterations: int = 200,
    log_interval: int = 10,
    learning_rate: float = 6e-4,
    batch_size: int = 1,
    epochs: int = 3,
    context_length: int = 4096,
    model_size: str = "65B",
    weight_decay: float = 1e-1,
):
    prepped_data_dir = prepare(
        input_file=data_dir,
        tokenizer_path=f"/scratch/data/models/llama-{model_size.lower()}/tokenizer.model"
    )
    train_data, val_data = load_datasets(prepped_data_dir)
    train_steps = calculate_training_steps(train_data, batch_size, epochs, context_length)

    print("Total number of training iterations: ", train_steps)

    ############################# TRAINING PREP #########################################
    auto_wrap_policy = partial(transformer_auto_wrap_policy, transformer_layer_cls={Block})
    strategy = FSDPStrategy(auto_wrap_policy=auto_wrap_policy, activation_checkpointing=Block, limit_all_gathers=True, cpu_offload=True, precision="16")

    fabric = L.Fabric(precision="bf16-mixed", strategy=strategy)

    # fabric.launch()
    fabric.seed_everything(1337 + fabric.global_rank)

    if fabric.global_rank == 0:
        os.makedirs(output_dir, exist_ok=True)
    ######################################################################################

    config = LLaMAConfig.from_name(model_size)
    config.block_size = context_length

    with fabric.init_module(empty_init=True):
        print("Loading model...")
        model = LLaMA(config).bfloat16()
        # Main machine, no need to load the weights again
        print("Loading checkpoint...")
        checkpoint = torch.load(f'/scratch/data/models/llama-{model_size.lower()}/lit-llama.pth')
        print("Loading state dict...")
        model.load_state_dict(checkpoint, strict=False)

    fabric.barrier()
    print("Setting up modules...")
    model = fabric.setup_module(model)

    print("Setting up the optimizers...")
    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate, weight_decay=weight_decay, betas=(beta1, beta2), foreach=False)
    optimizer = fabric.setup_optimizers(optimizer)

    loop(
        fabric,
        model,
        optimizer,
        train_data,
        val_data,
        eval_interval,
        eval_iterations,
        output_dir,
        batch_size,
        log_interval,
        train_steps
    )

############################### MAIN TRAINING LOOP ########################################
"""The training loop.

Loosely based on the nanoGPT implementation: https://github.com/karpathy/nanoGPT.
"""
def loop(
    fabric,
    model,
    optimizer,
    train_data,
    val_data,
    eval_interval,
    eval_iterations,
    output_dir,
    batch_size,
    log_interval,
    train_steps  
):
    iter_num = 0

    print("Beginning training loop ---------->")
    while True:
        # TODO: add learning rate scheduling

        # evaluate the loss on train/val sets and write checkpoints
        if iter_num > 0 and iter_num % eval_interval == 0:
            val_loss = validate(fabric, model, val_data, eval_iterations)
            fabric.print(f"step {iter_num}: val loss {val_loss:.4f}")
            fabric.print(f"Saving checkpoint to {output_dir}")
            save_model_checkpoint(fabric, model, os.path.join(output_dir, f"iter-{iter_num:06d}-ckpt.pth"))

        t0 = time.time()

        input_ids, targets = get_batch(
            fabric,
            train_data,
            block_size=model.config.block_size,  # type: ignore[union-attr,arg-type]
            batch_size=batch_size
        )
        logits = model(input_ids)
        loss = torch.nn.functional.cross_entropy(logits.view(-1, logits.size(-1)), targets.view(-1), ignore_index=-1)

        fabric.backward(loss)

        # TODO: Gradient clipping
        # if grad_clip != 0.0:
        #     fabric.clip_gradients(model, optimizer, max_norm=grad_clip)

        optimizer.step()
        optimizer.zero_grad()

        dt = time.time() - t0
        if iter_num % log_interval == 0:
            fabric.print(f"iter {iter_num}: loss {loss.item():.4f}, time: {dt*1000:.2f}ms")
        iter_num += 1

        if iter_num > train_steps:
            save_model_checkpoint(fabric, model, os.path.join(output_dir, f"iter-{iter_num:06d}-ckpt.pth"))
            break
```

I run this script using:
```
lightning run model \
    --main-port=29415 \
    --node-rank=0 \
    --main-address=IP_ADDRESS \
    --accelerator=cuda \
    --devices=8 \
    --num-nodes=2 \
    train.py \
        --ARGS...
```
on two separate machines with a different node rank. I'm seeing two bugs, although it just might be me not seeing something
1. When training the 30B model, the weights get initialized properly on the master Node, but on the child node, all the weights are dumped into GPU 1, which gets an OOM. No other GPUs are used. This happens in the Llama(config) step.
2. When training the 65B model, the weights are initialized on node 1, but there are only 8 GPUs with 80gb each (640GB) which is not enough to load llama, and I get an OOM again. I need to shard this across both the GPUs and the Nodes, and it seems to load a copy of llama on each node rather than one llama copy and shard across the nodes.

Is there something I'm missing or is this a bug? Thanks!

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
See above code
```

### Error messages and logs

```
CUDA oom
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

cc @carmocca @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #16448 
Fixes #18008

```py
fabric = Fabric(strategy="fsdp")

with fabric.init_module(empty_init=True):
    model = Model() # the model is now on the meta device (no memory occupied)
...
# materialization and param init and sharding happens here:
model = fabric.setup(model)

# you can now start training
train(model)
```

This allows you to instantiate very large models that wouldn't fit in memory (either CPU or GPU) as fast as possible. No memory for weights get allocated, neither in CPU nor GPU memory and parameters are materialized/initialized with random weights directly at the time the model gets wrapped and sharded in `Fabric.setup()`.

**Notes:**
- This new feature is possible thanks to https://github.com/pytorch/pytorch/issues/104187 in PyTorch 2.1 nightly.
Requirement: Your submodules define a `reset_parameters()` method that can be called to init the params. This is the case for all built-in PyTorch layers. If you have a custom layer, you'd have to add that method.
- I have tested this in the lit-gpt repo with the full finetuning and the pretraining scripts.
- Documentation will be updated in a follow-up. Reason: I'd like to send a PR to lit-gpt first to apply this in all models, making sure there are no unforeseen caveats.

**Limitation:**

Since the model is put on the meta device, you can't reference the parameters of that model in an optimizer like so:
```py
with fabric.init_module(empty_init=True):
	model = ...
optimizer = Adam(model.parameters())  # references meta-device parameters
model, optimizer = fabric.setup(model, optimizer)  # error
```

The user has to change the code to set up the model first, then create the optimizer referencing the FSDP parameters:
```py
with fabric.init_module(empty_init=True):
	model = ...
model = fabric.setup(model)  # set up model first
optimizer = Adam(model.parameters())  # references real parameters
optimizer = fabric.setup_optimizers(optimizer)
```
We have checks for this and explain it to the user in an error message. This will also be documented (see note above) and we will find a way to lift this limitation in the future. 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by carmocca — COMMENTED** [review]
Great work here. Are you planning to port this to the Trainer? If not, we should open an issue. We should do it before 2.1

File: src/lightning/fabric/strategies/fsdp.py [hunk 1]
```diff
@@ -317,9 +322,16 @@ def module_to_device(self, module: Module) -> None:
...
     def module_init_context(self, empty_init: Optional[bool] = None) -> Generator[None, None, None]:
-        # TODO: Use the meta device and reset parameters after https://github.com/pytorch/pytorch/issues/90465
-        # is resolved. For now, the module will get moved to the device in `setup_module`.
-        empty_init_context = _EmptyInit(enabled=bool(empty_init)) if _TORCH_GREATER_EQUAL_1_13 else nullcontext()
+        empty_init_context: Union[torch.device, _EmptyInit, nullcontext]
+        if _TORCH_GREATER_EQUAL_2_1 and empty_init:
+            # Materialization happens in `setup`. When modules get wrapped by FSDP, the sequence of operations is:
+            # 1) materialize module 2) call `reset_parameters()` 3) shard the module.
+            # These operations are applied to each submodule 'bottom up' in the module hierarchy.
+            empty_init_context = torch.device("meta")
```
Comment (carmocca) [inline src/lightning/fabric/strategies/fsdp.py:330]: Doesn't this break the loading of checkpoints for models that haven't been FSDP wrapped with `setup` yet? For instance: https://github.com/Lightning-AI/lit-gpt/blob/1900b80424825cb221af0b63d19dd33b027d9aff/generate/base.py#L145-L151

Wouldn't `load_state_dict` need `assign=True` now? https://github.com/pytorch/pytorch/issues/96161#issuecomment-1655690355
